File: app/src/main/java/com/yeraygarcia/recipes/database/dao/BaseDao.java
package com.yeraygarcia.recipes.database.dao;

import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Transaction;
import android.arch.persistence.room.Update;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static android.arch.persistence.room.OnConflictStrategy.IGNORE;

abstract class BaseDao<T> {

    @Delete
    public abstract int delete(T entity);

    @Delete
    public abstract int delete(List<T> entities);

    @Insert
    public abstract long insert(T entity);

    @Insert
    public abstract List<Long> insert(List<T> entities);

    @Insert(onConflict = IGNORE)
    abstract List<Long> insertOrIgnore(List<T> entities);

    @Update
    public abstract void update(T entity);

    @Update
    public abstract void update(List<T> entities);

    @Transaction
    public void upsert(T entity) {
        upsert(Collections.singletonList(entity));
    }

    @Transaction
    public void upsert(List<T> entities) {
        List<Long> insertResult = insertOrIgnore(entities);
        List<T> updateList = new ArrayList<>();

        for (int i = 0; i < insertResult.size(); i++) {
            if (insertResult.get(i) == -1) {
                updateList.add(entities.get(i));
            }
        }

        if (!updateList.isEmpty()) {
            update(updateList);
        }
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/database/dao/BaseDao.kt
package com.yeraygarcia.recipes.database.dao

import android.arch.persistence.room.Delete
import android.arch.persistence.room.Insert
import android.arch.persistence.room.OnConflictStrategy.IGNORE
import android.arch.persistence.room.Transaction
import android.arch.persistence.room.Update
import java.util.*

abstract class BaseDao<T> {

    @Delete
    abstract fun delete(entity: T): Int

    @Delete
    abstract fun delete(entities: List<T>): Int

    @Insert
    abstract fun insert(entity: T): Long

    @Insert
    abstract fun insert(entities: List<T>): List<Long>

    @Insert(onConflict = IGNORE)
    internal abstract fun insertOrIgnore(entities: List<T>): List<Long>

    @Update
    abstract fun update(entity: T)

    @Update
    abstract fun update(entities: List<T>)

    @Transaction
    open fun upsert(entity: T) {
        upsert(listOf(entity))
    }

    @Transaction
    open fun upsert(entities: List<T>) {
        val insertResult = insertOrIgnore(entities)
        val updateList = ArrayList<T>()

        for (i in insertResult.indices) {
            if (insertResult[i] == -1L) {
                updateList.add(entities[i])
            }
        }

        if (!updateList.isEmpty()) {
            update(updateList)
        }
    }
}
